rclpy/src/rclpy/events_executor/timers_manager.cpp

```diff
@@ -0,0 +1,335 @@
+{
+  auto cb = reinterpret_cast<const TimerResetCallbackT *>(user_data);
+  (*cb)();
+}
+
+}  // namespace
+
+/// Manages a single clock source, and all timers operating on that source.  All methods (including
+/// construction and destruction) are assumed to happen on the thread running the provided asio
+/// executor.
+class RclTimersManager::ClockManager
+{
+public:
+  ClockManager(const asio::any_io_executor & executor, rcl_clock_t * clock)
+  : executor_(executor), clock_(clock)
+  {
+    // Need to establish a clock jump callback so we can tell when debug time is updated.
+    rcl_jump_threshold_t threshold{.on_clock_change = true, .min_forward = 1, .min_backward = -1};
+    // Note, this callback could happen on any thread
+    jump_cb_ = [this](const rcl_time_jump_t * time_jump) {
+        bool on_debug{};
+        switch (time_jump->clock_change) {
+          case RCL_ROS_TIME_NO_CHANGE:
+          case RCL_ROS_TIME_ACTIVATED:
+            on_debug = true;
+            break;
+          case RCL_ROS_TIME_DEACTIVATED:
+          case RCL_SYSTEM_TIME_NO_CHANGE:
+            on_debug = false;
+            break;
+        }
+        asio::post(executor_, std::bind(&ClockManager::HandleJump, this, on_debug));
+      };
+    if (
+      RCL_RET_OK !=
+      rcl_clock_add_jump_callback(clock_, threshold, RclClockJumpTrampoline, &jump_cb_))
+    {
+      throw std::runtime_error(
+        std::string("Failed to set RCL clock jump callback: ") + rcl_get_error_string().str);
+    }
+
+    // This isn't necessary yet but every timer will eventually depend on it.  Again, this could
+    // happen on any thread.
+    reset_cb_ = [this]() {asio::post(executor_, std::bind(&ClockManager::UpdateTimers, this));};
+
+    // Initialize which timebase we're on
+    if (clock_->type == RCL_ROS_TIME) {
+      if (RCL_RET_OK != rcl_is_enabled_ros_time_override(clock_, &on_debug_time_)) {
+        throw std::runtime_error(
+          std::string("Failed to get RCL clock override state: ") + rcl_get_error_string().str);
+      }
+    }
+  }
+
+  ~ClockManager()
+  {
+    if (RCL_RET_OK != rcl_clock_remove_jump_callback(clock_, RclClockJumpTrampoline, &jump_cb_)) {
+      py::gil_scoped_acquire gil_acquire;
+      py::print(
+        std::string("Failed to remove RCL clock jump callback: ") + rcl_get_error_string().str);
+    }
+    while (!timers_.empty()) {
+      RemoveTimer(timers_.begin()->first);
+    }
+  }
+
+  bool empty() const {return timers_.empty();}
+
+  void AddTimer(rcl_timer_t * timer, std::function<void()> ready_callback)
+  {
+    // All timers have the same reset callback
+    if (RCL_RET_OK != rcl_timer_set_on_reset_callback(timer, RclTimerResetTrampoline, &reset_cb_)) {
+      throw std::runtime_error(
+        std::string("Failed to set timer reset callback: ") + rcl_get_error_string().str);
+    }
+    timers_[timer] = ready_callback;
+    if (timers_.size() == WARN_TIMERS_COUNT) {
```

Comment:
```suggestion
    if (timers_.size() >= WARN_TIMERS_COUNT) {
```

Reply:
So with this change, you'd get the warning re-printed with each additional timer beyond 8, rather than just once, which I was trying to avoid.  Granted, nothing stops you from creating seven timers, and then repeatedly creating and destroying the 8th, and getting the warning repeatedly anyway.

Reply:
I would consider more warning a better behavior. The rational being, that nodes tend to be spammy with their output, and you could easily miss the warning. Also the management should get worse with every additional added timer, therefore I would say its worth printing an additional warning.